cnd/src/storage.rs

```diff
@@ -542,109 +654,487 @@ impl
+                },
+                secret_hash: secret_hash
+                    .ok_or(db::Error::SecretHashNotSet)?
+                    .secret_hash
+                    .0,
+            }),
+            _ => Ok(http_api::BobSwap::Created {
+                alpha_created: hbit_asset,
+                beta_created: herc20_asset,
+            }),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Load<db::Swap> for Storage {
+    async fn load(&self, swap_id: LocalSwapId) -> anyhow::Result<db::Swap> {
+        #[derive(QueryableByName)]
+        struct Result {
+            #[sql_type = "sql_types::Text"]
+            role: Text<Role>,
+            #[sql_type = "sql_types::Text"]
+            alpha_protocol: Text<Protocol>,
+            #[sql_type = "sql_types::Text"]
+            beta_protocol: Text<Protocol>,
+        }
+
+        let Result { role, alpha_protocol, beta_protocol } = self.db.do_in_transaction(|connection| {
+            // Here is how this works:
+            // - COALESCE selects the first non-null value from a list of values
+            // - We use 3 sub-selects to select a static value (i.e. 'halight', etc) if that particular child table has a row with a foreign key to the parent table
+            // - We do this two times, once where we limit the results to rows that have `ledger` set to `Alpha` and once where `ledger` is set to `Beta`
+            //
+            // The result is a view with 3 columns: `role`, `alpha_protocol` and `beta_protocol` where the `*_protocol` columns have one of the values `halight`, `herc20` or `hbit`
+            diesel::sql_query(
+                r#"
+                SELECT
+                    role,
+                    COALESCE(
+                       (SELECT 'halight' from halights where halights.swap_id = swaps.id and halights.side = 'Alpha'),
+                       (SELECT 'herc20' from herc20s where herc20s.swap_id = swaps.id and herc20s.side = 'Alpha'),
+                       (SELECT 'hbit' from hbits where hbits.swap_id = swaps.id and hbits.side = 'Alpha')
+                    ) as alpha_protocol,
+                    COALESCE(
+                       (SELECT 'halight' from halights where halights.swap_id = swaps.id and halights.side = 'Beta'),
+                       (SELECT 'herc20' from herc20s where herc20s.swap_id = swaps.id and herc20s.side = 'Beta'),
+                       (SELECT 'hbit' from hbits where hbits.swap_id = swaps.id and hbits.side = 'Beta')
+                    ) as beta_protocol
+                from swaps
+                    where local_swap_id = ?
+            "#,
+            )
+                .bind::<sql_types::Text, _>(Text(swap_id))
+                .get_result(connection)
+        }).await.context(db::Error::SwapNotFound)?;
+
+        Ok(db::Swap {
+            role: role.0,
+            alpha: alpha_protocol.0,
+            beta: beta_protocol.0,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl Load<halight::Params> for Storage {
```

Comment:
Can we have a ticket for deciding on:

- Either loading these individually but with an expectation of which "side" we expect this to be
- Or only loading params in pairs where we assert that what we find matches the side the user requested it on

Reply:
Done with #2781, but `to-refine` unless we tackle it very soon.